psi4/src/psi4/libmints/electricfield.h

```diff
@@ -68,6 +71,49 @@ class ElectricFieldInt : public OneBodyAOInt {
 
     static Vector3 nuclear_contribution(const Vector3 &origin, std::shared_ptr<Molecule> mol);
     static SharedMatrix nuclear_contribution_to_gradient(const Vector3 &origin, std::shared_ptr<Molecule> mol);
+
+    template <typename ContractionFunctor>
+    void compute_with_functor(ContractionFunctor &functor, SharedMatrix coords);
+};
+
+class ContractOverDipolesFunctor {
+   protected:
+    /// Pointer to the resulting operator matrix
+    double **pF_;
+    /// The array of dipoles
+    double **dipoles_;
+
+   public:
+    ContractOverDipolesFunctor(SharedMatrix dipoles, SharedMatrix F) : pF_(F->pointer()), dipoles_(dipoles->pointer()) {
+        if (F->rowdim() != F->coldim()) throw PSIEXCEPTION("Invalid Fock matrix in ContractOverCharges");
+        if (dipoles->coldim() != 3) throw PSIEXCEPTION("Dipole matrix must have 3 columns.");
+        int nbf = F->rowdim();
+        std::fill_n(pF_[0], nbf * nbf, 0.0);
+    }
+
+    void operator()(int bf1, int bf2, int center, double integralx, double integraly, double integralz) {
+        pF_[bf1][bf2] +=
+            integralx * dipoles_[center][0] + integraly * dipoles_[center][1] + integralz * dipoles_[center][2];
+    }
+};
+
+class ContractOverDensityFieldFunctor {
+   protected:
+    /// The array of the density matrix
+    double **pD_;
+    /// The array of electric fields
+    double **field_;
+
+   public:
+    ContractOverDensityFieldFunctor(SharedMatrix field, SharedMatrix D) : pD_(D->pointer()), field_(field->pointer()) {
+        if (D->rowdim() != D->coldim()) throw PSIEXCEPTION("Invalid density matrix in ContractOverDensityFieldFunctor");
+        if (field->coldim() != 3) throw PSIEXCEPTION("Field matrix must have 3 columns.");
```

Comment:
Here you didn't use `std::fill_n` on `field_` which is how I think it should be.

Reply:
Done.